funderfinder/sources/numfocus.py

```diff
@@ -12,42 +15,60 @@
 """
 
 
-def get_funding_stats(search_params: dict) -> dict:
-    """
-    Determines whether a project is sponsored or affiliated with NumFOCUS based on our scraped dataset and
-    a project name, slug, or github owner and repo name
-    :param search_params: Dict of user-provided metadata that we can use to match a numfocus project
-    :return: Dict of funding stats
-    """
-    is_affiliated = False
-    with open(
-        os.path.join(
-            os.path.dirname(os.path.abspath(__file__)), "..", "data", "numfocus.jsonl"
+class NumFocusFinder(Finder):
+    name = "NumFOCUS"
+
+    @staticmethod
+    def get_funding_stats(search_params: dict) -> Union[dict, None]:
+        """
+        Determines whether a project is sponsored or affiliated with NumFOCUS based on our scraped dataset and
+        a project name, slug, or github owner and repo name
+        :param search_params: Dict of user-provided metadata that we can use to match a numfocus project
+        :return: Dict of funding stats
+        """
+        is_affiliated = False
+        with open(
+            os.path.join(
+                os.path.dirname(os.path.abspath(__file__)),
+                "..",
+                "data",
+                "numfocus.jsonl",
+            )
+        ) as f:
+            for line in f:
+                project_metadata = json.loads(line.strip())
+                for key in search_params:
+                    # In this block, we iterate through metadata fields provided by the user to match a numfocus project.
+                    # Some of these fields may be null either in the user-provided input (`search_params`), or in the
+                    # metadata we have for the current numfocus project (`project_metadata`). If either of these values
+                    # are null, run no further checks
+                    if not (project_metadata[key] and search_params[key]):
+                        continue
+                    is_affiliated |= (
+                        project_metadata[key].lower() == search_params[key].lower()
+                    )
+                    # In some cases the NumFOCUS affiliation is at the GitHub organization level rather than at the repo
+                    # level. So also allow match on repo owner
+                    if key == "github_name":
+                        owner = search_params[key].split("/")[0].lower()
+                        is_affiliated |= project_metadata[key].lower() == owner
+        if is_affiliated:
+            return {
+                "is_funded": True,
+            }
+
+    def run(
+        self,
+        gh_project_slug: Union[str, None] = None,
```

Comment:
I don't understand these `Union[str, None] = None` blocks. What's going on?

Reply:
Oh, they're just [type hints](https://docs.python.org/3/library/typing.html) with a default value. In this case, the hint is saying that gh_project_slug will either be None or a string, and the default value is None.

Reply:
Ah, thank you!